hathor/indexes/manager.py

```diff
@@ -66,23 +110,82 @@ def enable_tokens_index(self) -> None:
         """Enable tokens index. It does nothing if it has already been enabled."""
         raise NotImplementedError
 
-    def _manually_initialize_tips_indexes(self, tx_storage: 'TransactionStorage') -> None:
-        """ Initialize the tips indexes, populating them from a tx_storage that is otherwise complete.
-
-        XXX: this method requires timestamp indexes to be complete and up-to-date with the rest of the database
-        XXX: this method is not yet being used
+    def _manually_initialize(self, tx_storage: 'TransactionStorage') -> None:
+        """ Initialize the indexes, checking the indexes that need initialization, and the optimal iterator to use.
         """
-        for tx in tx_storage._topological_sort_timestamp_index():
-            tx_meta = tx.get_metadata()
-            if not tx_meta.validation.is_final():
-                continue
+        from hathor.transaction.genesis import BLOCK_GENESIS
 
-            self.all_tips.add_tx(tx)
+        db_last_started_at = tx_storage.get_last_started_at()
+        sorted_all_db_name = self.sorted_all.get_db_name()
+        if sorted_all_db_name is None:
+            can_use_timestamp_index = False
+        else:
+            sorted_all_index_last_started_at = tx_storage.get_index_last_started_at(sorted_all_db_name)
+            can_use_timestamp_index = db_last_started_at == sorted_all_index_last_started_at
+
+        indexes_to_init: List[Tuple[_IndexFilter, BaseIndex]] = []
+        for index_filter, index in self._iter_all_indexes_with_filter():
+            index_db_name = index.get_db_name()
+            if index_db_name is None:
+                indexes_to_init.append((index_filter, index))
+                continue
+            index_last_started_at = tx_storage.get_index_last_started_at(index_db_name)
+            if db_last_started_at != index_last_started_at:
+                indexes_to_init.append((index_filter, index))
 
+        if indexes_to_init:
+            self.log.debug('there are indexes that need initialization',
+                           indexes_to_init=[i for _, i in indexes_to_init])
+        else:
+            self.log.debug('there are no indexes that need initialization')
+
+        iter_tx: Iterator[BaseTransaction]
+        if can_use_timestamp_index:
+            iter_tx = tx_storage._topological_sort_timestamp_index()
+        # XXX: this mostly exist because some tests will break when not using a caching-storage because of metadata
+        #      objects being duplicated which make the topological metadata iterator not work
+        elif tx_storage.is_caching_storage():
```

Comment:
Why is this the criteria to use `_topological_sort_metadata`?

Reply:
It's not like this **is the criteria**, but as I explained in the comment above, some tests will break when using `_topological_sort_metadata` here, it's tests that do not use a caching-storage, and it's related to the sharing of metadata objects. This is _probably_ a bug on how we handle `TransactionMetadata` objects, but I'd rather not deal with this bug right now, since _apparently_ it only affects tests. But I do think this deserves at least an issue to investigate later.

Reply:
I just don't think we should add a criteria to pass the tests. Should we add a parameter to IndexesManager to pick which topological sort it will use?

We should also create an issue to investigate duplicate metadata. Please, add the steps to generate the issue there.

Reply:
I think I got what you mean, now it might make sense to move that to the storage, since it's related to the storage I think it's better to have a flag/toggle there, which would be temporarily used on tests that fail when using `_topological_sort_metadata`. I'd say the storage "owns" the indexes so it might make more sense to have an iterator selector there anyway.

Reply:
Have you done it?

Reply:
Yes.

Reply:
Just reviewed the code. Thanks!